drivers/flash/spi_nor.c

```diff
@@ -0,0 +1,510 @@
+	};
+
+	const struct spi_buf_set rx_set = {
+		.buffers = spi_buf,
+		.count = 2
+	};
+
+	if (is_write) {
+		return spi_write(driver_data->spi, &driver_data->spi_cfg, &tx_set);
+	}
+
+	return spi_transceive(driver_data->spi, &driver_data->spi_cfg, &tx_set, &rx_set);
+}
+#define spi_nor_cmd_read(dev, opcode, dest, length) \
+	spi_nor_access(dev, opcode, false, 0, dest, length, false)
+#define spi_nor_cmd_addr_read(dev, opcode, addr, dest, length) \
+	spi_nor_access(dev, opcode, true, addr, dest, length, false)
+#define spi_nor_cmd_write(dev, opcode) \
+	spi_nor_access(dev, opcode, false, 0, NULL, 0, true)
+#define spi_nor_cmd_addr_write(dev, opcode, addr, src, length) \
+	spi_nor_access(dev, opcode, true, addr, src, length, true)
+
+/**
+ * @brief Retrieve the Flash JEDEC ID and compare it with the one expected
+ *
+ * @param dev The device structure
+ * @param flash_id The flash info structure which contains the expected JEDEC ID
+ * @return 0 on success, negative errno code otherwise
+ */
+static inline int spi_nor_read_id(struct device *dev, const struct flash_info *const flash_id)
+{
+	u8_t buf[SPI_NOR_MAX_ID_LEN];
+
+	if (spi_nor_cmd_read(dev, SPI_NOR_CMD_RDID, buf, SPI_NOR_MAX_ID_LEN) != 0) {
+		return -EIO;
+	}
+
+	if (memcmp(flash_id->id, buf, SPI_NOR_MAX_ID_LEN) != 0) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Wait until the flash is ready
+ *
+ * @param dev The device structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static inline int spi_nor_wait_until_ready(struct device *dev)
+{
+	int ret;
+	u8_t reg;
+
+	do {
+		ret = spi_nor_cmd_read(dev, SPI_NOR_CMD_RDSR, &reg, 1);
+	} while (!ret && (reg & SPI_NOR_WIP_BIT));
+
+	return ret;
+}
+
+/**
+ * @brief Read data from the flash
+ *
+ * @param dev The flash device structure
+ * @param addr The address to read the value from
+ * @param dest The buffer to store the read value
+ * @param size The size of the buffer
+ * @return number of byte read on success, negative errno code otherwise
+ */
+static int spi_nor_read(struct device *dev, off_t addr, void *dest,
+			size_t size)
+{
+	struct spi_nor_data *const driver_data = dev->driver_data;
+	struct spi_nor_parameters *const params = &driver_data->params;
+	const void *const p_dest = dest;
+	int ret;
+	int to_read;
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || (addr + size) > params->flash_size) {
+		return -EINVAL;
+	}
+
+	/* start address should be aligned on a page size */
+	if ((addr & (params->page_size - 1)) != 0) {
+		return -EINVAL;
+	}
+
+	SYNC_LOCK();
+
+	spi_nor_wait_until_ready(dev);
+
+	while (size) {
+		to_read = size;
+		if (size > params->page_size) {
+			to_read = params->page_size;
+		}
+
+		ret = spi_nor_cmd_addr_read(dev, SPI_NOR_CMD_READ, addr, dest, to_read);
+		if (ret != 0) {
+			goto err;
+		}
+		size -= to_read;
+		addr += to_read;
+		dest += to_read;
+	}
+
+err:
+	SYNC_UNLOCK();
+
+	return (dest - p_dest);
+}
+
+/**
+ * @brief Write data to the flash
```

Comment:
You don't need to repeat the documentation which is the same as flash API. So remove it for write, read, etc...